src/hotspot/share/services/heapDumper.cpp

```diff
@@ -608,20 +594,294 @@ void DumpWriter::start_sub_record(u1 tag, u4 len) {
+ private:
+  CompressionBackend _backend; // Does the actual writing.
+ protected:
+  virtual void flush(bool force = false);
+
+ public:
+  // Takes ownership of the writer and compressor.
+  DumpWriter(AbstractWriter* writer, AbstractCompressor* compressor);
+
+  // total number of bytes written to the disk
+  virtual julong bytes_written() const          { return (julong) _backend.get_written(); }
+
+  virtual char const* error() const             { return _backend.error(); }
+
+  // Called by threads used for parallel writing.
+  void writer_loop()                    { _backend.thread_loop(false); }
+  // Called when finish to release the threads.
+  virtual void deactivate()             { flush(); _backend.deactivate(); }
+  // Get the backend pointer, used by parallel dump writer.
+  CompressionBackend* backend_ptr() { return &_backend; }
+
+};
+
+// Check for error after constructing the object and destroy it in case of an error.
+DumpWriter::DumpWriter(AbstractWriter* writer, AbstractCompressor* compressor) :
+  AbstractDumpWriter(),
+  _backend(writer, compressor, io_buffer_max_size, io_buffer_max_waste) {
+  flush();
+}
+
+// flush any buffered bytes to the file
+void DumpWriter::flush(bool force) {
+  _backend.get_new_buffer(&_buffer, &_pos, &_size, force);
+}
+
+// Buffer queue used for parallel dump.
+struct ParWriterBufferQueueElem {
+  char* _buffer;
+  size_t _used;
+  ParWriterBufferQueueElem* _next;
+};
+
+class ParWriterBufferQueue : public CHeapObj<mtInternal> {
+ private:
+  ParWriterBufferQueueElem* _head;
+  ParWriterBufferQueueElem* _tail;
+  uint _length;
+ public:
+  ParWriterBufferQueue() : _head(NULL), _tail(NULL), _length(0) { }
+
+  void enqueue(ParWriterBufferQueueElem* entry) {
+    if (_head == NULL) {
+      assert(is_empty() && _tail == NULL, "Sanity check");
+      _head = _tail = entry;
+    } else {
+      assert ((_tail->_next == NULL && _tail->_buffer != NULL), "Buffer queue is polluted");
+      _tail->_next = entry;
+      _tail = entry;
+    }
+    _length++;
+    assert(_tail->_next == NULL, "Bufer queue is polluted");
+  }
+
+  ParWriterBufferQueueElem* dequeue() {
+    if (_head == NULL)  return NULL;
+    ParWriterBufferQueueElem* entry = _head;
+    assert (entry->_buffer != NULL, "polluted buffer in writer list");
+    _head = entry->_next;
+    if (_head == NULL) {
+      _tail = NULL;
+    }
+    entry->_next = NULL;
+    _length--;
+    return entry;
+  }
+
+  bool is_empty() {
+    return _length == 0;
+  }
+
+  uint length() { return _length; }
+};
+
+// Support parallel heap dump.
+class ParDumpWriter : public AbstractDumpWriter {
+ private:
+  // Lock used to guarantee the integrity of multiple buffers writing.
+  static Monitor* _lock;
+  // Pointer of backend from global DumpWriter.
+  CompressionBackend* _backend_ptr;
+  char const * _err;
+  ParWriterBufferQueue* _buffer_queue;
+  size_t _internal_buffer_used;
+  char* _buffer_base;
+  bool _splited_data;
```

Comment:
`_split_data`. Split is one of those english verbs that doesn't follow the rules. "I split the log yesterday" is correct, not "splited".